download/download_generator.go

```diff
@@ -0,0 +1,106 @@
+package download
+
+import (
+	"fmt"
+
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+)
+
+//go:generate moq -out ../mocks/generate_downloads_mocks.go -pkg mocks . KafkaProducer GenerateDownloadsEvent
+
+var (
+	avroMarshalErr = "error while attempting to marshal generateDownloadsEvent to avro bytes"
+
+	datasetIDEmptyErr  = newGeneratorError(nil, "failed to generate full dataset download as dataset ID was empty")
+	instanceIDEmptyErr = newGeneratorError(nil, "failed to generate full dataset download as instance ID was empty")
+	editionEmptyErr    = newGeneratorError(nil, "failed to generate full dataset download as edition was empty")
+	versionEmptyErr    = newGeneratorError(nil, "failed to generate full dataset download as version was empty")
+)
+
+// KafkaProducer sends an outbound kafka message
+type KafkaProducer interface {
+	Output() chan []byte
+}
+
+// GenerateDownloadsEvent marshal the event into avro format
+type GenerateDownloadsEvent interface {
+	Marshal(s interface{}) ([]byte, error)
+}
+
+type generateDownloads struct {
+	FilterID   string `avro:"filter_output_id"`
+	InstanceID string `avro:"instance_id"`
+	DatasetID  string `avro:"dataset_id"`
+	Edition    string `avro:"edition"`
+	Version    string `avro:"version"`
+}
+
+// Generator kicks off a full dataset version download task
+type Generator struct {
+	Producer   KafkaProducer
+	Marshaller GenerateDownloadsEvent
+}
+
+// Generate the full file download files for the specified dataset/edition/version
+func (gen *Generator) Generate(datasetID string, instanceID string, edition string, version string) error {
+	if datasetID == "" {
+		return datasetIDEmptyErr
+	}
+	if instanceID == "" {
+		return instanceIDEmptyErr
+	}
+	if edition == "" {
+		return editionEmptyErr
+	}
+	if version == "" {
+		return versionEmptyErr
+	}
+
+	generateDownloads := generateDownloads{
+		FilterID:   "",
```

Comment:
I assume the avro schema expects a filter ID, even if it is an empty string. If the variable was not here it would not marshal unless we change the schema to make this an optional field.